### PR TITLE
Feat/frontend bugs v4

### PR DESCRIPTION
**Pull request by mostarz93**
- [x] Dont allow borrow if cratio <= minCratio
- [x] if no debt, cratio is infitity
- [x] Implement getInterestAccrued
- [x] setRates
- [x] max borrow takes into account drawn amount
- [x] auto populate claim func bytecode
- [x] collateral table accuracy

### CODE REVIEW

File: src/client/components/app/Transactions/BorrowCreditTx.tsx [hunk 1]
```diff
@@ -54,6 +54,12 @@ export const BorrowCreditTx: FC<BorrowCreditProps> = (props) => {
     _updatePosition();
   };
 
+  const getMax = (pos: CreditPosition): string => {
+    const maxBorrow = Number(pos.deposit) - Number(pos.principal);
```
Comment (hepnerthomas) [inline src/client/components/app/Transactions/BorrowCreditTx.tsx]: you should be doing math with `BigNumber`, not `Number` 

do this instead: 

   
```
  const amountAvailableToBorrow = normalizeAmount(
BigNumber.from(selectedPosition.deposit).sub(BigNumber.from(selectedPosition.principal)).toString(),
    BASE_DECIMALS
  );
```

This code is in my branch already: https://github.com/debtdao/marketplace-app/pull/111

File: src/core/store/modules/collateral/collateral.reducer.ts [hunk 2]
```diff
@@ -195,13 +196,22 @@ const collateralReducer = createReducer(collateralInitialState, (builder) => {
...
+    .addCase(
+      tradeable.fulfilled,
+      (state, { payload: { tokenAddressMap, spigotAddress, tokenPrices, tokenAddress, lineAddress } }) => {
+        const map = { [tokenAddress]: tokenAddressMap };
+        state.reservesMap = {
+          ...state.reservesMap,
+          [lineAddress]: { ...(state.reservesMap[lineAddress] || {}), ...map },
+        };
+        // update spigot revenue
+        console.log(spigotAddress);
```
Comment (hepnerthomas) [inline src/core/store/modules/collateral/collateral.reducer.ts:209]: remove

File: src/utils/bytesToName.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { Interface, ParamType } from '@ethersproject/abi';
+import { BytesLike } from 'ethers';
+
+// import ParamType type
+
+function bytesToName(funcBytes: BytesLike, contractABI: string): string {
+  const iface = new Interface(contractABI!);
+  let name: string = '';
+  console.log('Interface Log', iface.functions);
```
Comment (hepnerthomas) [inline src/utils/bytesToName.ts]: remove console.logs

File: src/utils/bytesToName.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  console.log(funcBytes);
+  for (const key in iface.functions) {
+    if (funcBytes === iface.getSighash(key)) {
+      name = iface.functions[key].name;
+    }
+  }
+  //   for (const key in iface.functions) {
```
Comment (hepnerthomas) [inline src/utils/bytesToName.ts]: remove dead code

File: src/utils/collateral.ts [hunk 5]
```diff
@@ -63,3 +66,70 @@ export const formatEscrowDeposit = (tokenInfo: TokenView): EscrowDeposit => {
...
+      unusedTokens: string;
+      ownerTokens: string;
+      operatorTokens: string;
+    };
+  },
+  tokenPrices: {
+    [address: string]: BigNumber;
+  }
+): AggregatedSpigot => {
+  // TODO: update collateralMap to remove operatorTokens from amount and usdValue of type === REVENUE
```
Comment (hepnerthomas) [inline src/utils/collateral.ts]: remove my old comments now that functionality is implemented

File: src/client/components/app/Transactions/SetRatesTx.tsx [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+      label: t('components.transaction.set-rates'),
+      onAction: setRates,
+      status: true,
+      disabled: false,
+      contrast: false,
+    },
+  ];
+
+  if (!selectedCredit) {
+    console.log('withdraw modal selected credit is undefined: ', selectedCredit);
```
Comment (hepnerthomas) [inline src/client/components/app/Transactions/SetRatesTx.tsx]: chang this console.log as its leftover from Withdraw component

File: src/client/components/app/Transactions/SetRatesTx.tsx [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    _updatePosition();
+  };
+
+  const onSelectedCreditLineChange = (addr: string): void => {
+    setSelectedCredit(addr);
+    _updatePosition();
+  };
+
+  const onRateChange = (type: string, amount: string): void => {
+    if (type === 'd') setDrate(amount);
```
Comment (hepnerthomas) [inline src/client/components/app/Transactions/SetRatesTx.tsx:119]: where are these types coming from? Can you rename them? single letter naming is pretty annoying when if you need to debug or troubleshoot later

File: src/client/components/app/Transactions/RepayPositionTx.tsx [hunk 8]
```diff
@@ -138,6 +138,12 @@ export const RepayPositionTx: FC<RepayPositionProps> = (props) => {
   const [isTrade, setIsTrade] = useState<boolean>(false);
   // TODO: replace usage in useFundsForClosing with appropriate value
 
+  useEffect(() => {
+    if (selectedPosition && selectedLine?.id) {
+      dispatch(LinesActions.getInterestAccrued({ contractAddress: selectedLine.id, id: selectedPosition.id }));
```
Comment (hepnerthomas) [inline src/client/components/app/Transactions/RepayPositionTx.tsx:153]: doesn't appear to be updating state via `lines.reducer` - will help with this today

**Review by hepnerthomas — CHANGES_REQUESTED** [review]

**Comment by hepnerthomas** [comment]
@mostarz93 Happened upon on this in `ReleaseCollateral.Tx`. Can you fix this line? 
```
        amountValue={String(10000000 * Number(targetTokenAmount))}
```

Why are you multiplying by 10000000? Can this be moved into constants?

Also, we should be using `BigNumber`, not `Number` to represent values.

**Comment by hepnerthomas** [comment]
> @mostarz93 Happened upon on this in `ReleaseCollateral.Tx`. Can you fix this line?
> 
> ```
>         amountValue={String(10000000 * Number(targetTokenAmount))}
> ```
> 
> Why are you multiplying by 10000000? Can this be moved into constants?
> 
> Also, we should be using `BigNumber`, not `Number` to represent values.

Looks like this is old code from 5 months ago. I'm not sure why it's doing this, but we have the same issue in `LiquidateBorrowerTx`, `AddCollateralTx`, and `WithdrawCredit`. 

I'm going to comment it out for now.

File: src/client/components/app/Transactions/SetRatesTx.tsx (hunk 7 above)
Comment (mostarz93) [reply src/client/components/app/Transactions/SetRatesTx.tsx:119]: I think this is how its done already for addCredit. I copied a lot of code for this func

File: src/core/store/modules/collateral/collateral.reducer.ts (hunk 2 above)
Comment (mostarz93) [reply src/core/store/modules/collateral/collateral.reducer.ts:209]: done

File: src/utils/bytesToName.ts (hunk 3 above)
Comment (mostarz93) [reply src/utils/bytesToName.ts]: done

File: src/utils/bytesToName.ts (hunk 4 above)
Comment (mostarz93) [reply src/utils/bytesToName.ts]: done

File: src/utils/collateral.ts (hunk 5 above)
Comment (mostarz93) [reply src/utils/collateral.ts]: done

File: src/client/components/app/Transactions/SetRatesTx.tsx (hunk 6 above)
Comment (mostarz93) [reply src/client/components/app/Transactions/SetRatesTx.tsx]: done

**Pull request closed without merging**